Update documentation to reflect k-NN FAISS AVX512 support

noahstaveley opened the pull request:
### Description
This change updates the documentation to reflect AVX512 hardware accelerator support that was added to speed up vector search and indexing. The k-NN index and settings pages under k-NN search are updated in this change. 

### Issues Resolved
Closes #8306 

### Version
2.18

Related to the PR: https://github.com/opensearch-project/k-NN/pull/2110

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

noahstaveley commented:
@akashsha1

akashsha1 commented:
@naveentatikonda - public documentation for AVX512 PR

kolchfa-aws commented:
@noahstaveley Could you please resolve DCO? Also, is this for versions 2.16 and above or for the next version (2.18)?

naveentatikonda commented:
@akashsha1 @noahstaveley I will try to review this PR by EOD. Thanks!

naveentatikonda commented on _search-plugins/knn/knn-index.md:
```diff
@@ -60,14 +60,19 @@ SIMD optimization is applicable only if the vector dimension is a multiple of 8.
 <!-- vale on -->
 
-For the x64 architecture, two different versions of the Faiss library are built and shipped with the artifact:
+For the x64 architecture, three different versions of the Faiss library are built and shipped with the artifact:
```
for the version, let us mention it as 2.18.0

noahstaveley commented:
> @noahstaveley Could you please resolve DCO? Also, is this for versions 2.16 and above or for the next version (2.18)?

DCO is resolved. This is for version 2.18 only, I updated the PR to reflect that. Thanks for confirming

naveentatikonda commented:
@noahstaveley Why are we seeing so many commits included in this PR. Can you rebase this PR again ?

noahstaveley commented:
> @noahstaveley Why are we seeing so many commits included in this PR. Can you rebase this PR again ?

@naveentatikonda  Yes sorry about that, I rebased and removed the unrelated commits.

naveentatikonda commented on _search-plugins/knn/knn-index.md:
```diff
@@ -60,14 +60,19 @@ SIMD optimization is applicable only if the vector dimension is a multiple of 8.
 If your hardware supports AVX2, the k-NN plugin loads the `libopensearchknn_faiss_avx2.so` library at runtime.
 
-To disable AVX2 and load the non-optimized Faiss library (`libopensearchknn_faiss.so`), specify the `knn.faiss.avx2.disabled` static setting as `true` in `opensearch.yml` (default is `false`). Note that to update a static setting, you must stop the cluster, change the setting, and restart the cluster. For more information, see [Static settings]({{site.url}}{{site.baseurl}}/install-and-configure/c ...
+If your hardware supports AVX512, the k-NN plugin loads the `libopensearchknn_faiss_avx512.so` library at runtime.
```
Can you move this line above line 69 because the order of preference is AVX512 > AVX2> Generic

naveentatikonda commented on _search-plugins/knn/knn-index.md:
```diff
@@ -60,14 +60,19 @@ SIMD optimization is applicable only if the vector dimension is a multiple of 8.
 - `libopensearchknn_faiss.so`: The non-optimized Faiss library without SIMD instructions. 
-- `libopensearchknn_faiss_avx2.so`: The Faiss library that contains AVX2 SIMD instructions. 
+- `libopensearchknn_faiss_avx2.so`: The Faiss library that contains AVX2 SIMD instructions.
+- `libopensearchknn_faiss_avx512.so`: The Faiss library that contains AVX512 SIMD instructions. 
```
Can you move avx512.so above avx2.so

naveentatikonda commented on _search-plugins/knn/settings.md:
```diff
@@ -27,6 +27,7 @@ Setting | Static/Dynamic | Default | Description
 `knn.model.index.number_of_replicas`| Dynamic | `1` | The number of replica shards to use for the model system index. Generally, in a multi-node cluster, this value should be at least 1 in order to increase stability.
 `knn.model.cache.size.limit` | Dynamic | `10%` |  The model cache limit cannot exceed 25% of the JVM heap.
 `knn.faiss.avx2.disabled` | Static | `false` | A static setting that specifies whether to disable the SIMD-based `libopensearchknn_faiss_avx2.so` library and load the non-optimized `libopensearchknn_faiss.so` library for the Faiss engine on machines with x64 architecture. For more information, see [SIMD optimization for the Faiss engine]({{site.url}}{{site.baseurl}}/search-plugins/knn/knn-index/# ...
+`knn.faiss.avx512.disabled` | Static | `false` | A static setting that specifies whether to disable the SIMD-based `libopensearchknn_faiss_avx512.so` library and load the non-optimized `libopensearchknn_faiss.so` library for the Faiss engine on machines with x64 architecture. For more information, see [SIMD optimization for the Faiss engine]({{site.url}}{{site.baseurl}}/search-plugins/knn/knn-ind ...
```
```suggestion
`knn.faiss.avx512.disabled` | Static | `false` | A static setting that specifies whether to disable the SIMD-based `libopensearchknn_faiss_avx512.so` library and load the `libopensearchknn_faiss_avx2.so` or the non-optimized `libopensearchknn_faiss.so` library for the Faiss engine on machines with x64 architecture. For more information, see [SIMD optimization for the Faiss engine]({{site.url}}{{site.baseurl}}/search-plugins/knn/knn-index/#simd-optimization-for-the-faiss-engine).
```

naveentatikonda commented on _search-plugins/knn/knn-index.md:
```diff
@@ -60,14 +60,19 @@ SIMD optimization is applicable only if the vector dimension is a multiple of 8.
-To disable AVX2 and load the non-optimized Faiss library (`libopensearchknn_faiss.so`), specify the `knn.faiss.avx2.disabled` static setting as `true` in `opensearch.yml` (default is `false`). Note that to update a static setting, you must stop the cluster, change the setting, and restart the cluster. For more information, see [Static settings]({{site.url}}{{site.baseurl}}/install-and-configure/c ...
+If your hardware supports AVX512, the k-NN plugin loads the `libopensearchknn_faiss_avx512.so` library at runtime.
+
+To disable AVX2 and AVX512 and load the non-optimized Faiss library (`libopensearchknn_faiss.so`), specify the `knn.faiss.avx2.disabled` and `knn.faiss.avx512.disabled` static settings as `true` in `opensearch.yml` (defaults are `false`). 
```
```suggestion
To disable AVX512 and AVX2 and load the non-optimized Faiss library (`libopensearchknn_faiss.so`), specify the `knn.faiss.avx2.disabled` and `knn.faiss.avx512.disabled` static settings as `true` in `opensearch.yml` (defaults are `false`). 
```

naveentatikonda commented on _search-plugins/knn/knn-index.md:
```diff
@@ -60,14 +60,19 @@ SIMD optimization is applicable only if the vector dimension is a multiple of 8.
+
+If your hardware supports AVX512, the k-NN plugin loads the `libopensearchknn_faiss_avx512.so` library at runtime.
 
 If your hardware supports AVX2, the k-NN plugin loads the `libopensearchknn_faiss_avx2.so` library at runtime.
```
```suggestion
If your hardware supports AVX2 but doesn't support AVX512, the k-NN plugin loads the `libopensearchknn_faiss_avx2.so` library at runtime.
```

naveentatikonda commented on _search-plugins/knn/knn-index.md:
```diff
@@ -60,14 +60,19 @@ SIMD optimization is applicable only if the vector dimension is a multiple of 8.
 <!-- vale on -->
 
-For the x64 architecture, two different versions of the Faiss library are built and shipped with the artifact:
+For the x64 architecture, three different versions of the Faiss library are built and shipped with the artifact:
```
Shall we add a note showing the order of performance avx512>avx2>generic ?

noahstaveley replied:
I added a note specifying order of performance below the list of libraries (line 69). Let me know if you think it would be better directly after line 63 instead.

naveentatikonda replied:
It needs to be added as a note. @kolchfa-aws can you help to fix it when you are reviewing it later

kolchfa-aws replied:
Sure, will do.

naveentatikonda commented:
@noahstaveley Thanks for raising this PR. @kolchfa-aws can you please review it. Thanks!

kolchfa-aws commented on _search-plugins/knn/knn-index.md:
```diff
@@ -60,14 +60,19 @@ SIMD optimization is applicable only if the vector dimension is a multiple of 8.
+- `libopensearchknn_faiss_avx2.so`: The Faiss library that contains AVX2 SIMD instructions.
 
-If your hardware supports AVX2, the k-NN plugin loads the `libopensearchknn_faiss_avx2.so` library at runtime.
+The Faiss library order of performance is AVX512, AVX2, generic.
```
```suggestion
When using the Faiss library, the performance ranking is: AVX512 > AVX2 > no optimization.
{: .note }

```

kolchfa-aws commented on _search-plugins/knn/knn-index.md:
```diff
@@ -60,14 +60,19 @@ SIMD optimization is applicable only if the vector dimension is a multiple of 8.
 
-If your hardware supports AVX2, the k-NN plugin loads the `libopensearchknn_faiss_avx2.so` library at runtime.
+The Faiss library order of performance is AVX512, AVX2, generic.
+If your hardware supports AVX512, the k-NN plugin loads the `libopensearchknn_faiss_avx512.so` library at runtime.
```
```suggestion
If your hardware supports AVX512, the k-NN plugin loads the `libopensearchknn_faiss_avx512.so` library at runtime.

```

kolchfa-aws commented on _search-plugins/knn/knn-index.md:
```diff
@@ -60,14 +60,19 @@ SIMD optimization is applicable only if the vector dimension is a multiple of 8.
+If your hardware supports AVX2 but doesn't support AVX512, the k-NN plugin loads the `libopensearchknn_faiss_avx2.so` library at runtime.
 
-To disable AVX2 and load the non-optimized Faiss library (`libopensearchknn_faiss.so`), specify the `knn.faiss.avx2.disabled` static setting as `true` in `opensearch.yml` (default is `false`). Note that to update a static setting, you must stop the cluster, change the setting, and restart the cluster. For more information, see [Static settings]({{site.url}}{{site.baseurl}}/install-and-configure/c ...
+To disable AVX512 and AVX2 and load the non-optimized Faiss library (`libopensearchknn_faiss.so`), specify the `knn.faiss.avx512.disabled` and `knn.faiss.avx2.disabled` static settings as `true` in `opensearch.yml` (defaults are `false`).
```
```suggestion
To disable the AVX512 and AVX2 SIMD instructions and load the non-optimized Faiss library (`libopensearchknn_faiss.so`), specify the `knn.faiss.avx512.disabled` and `knn.faiss.avx2.disabled` static settings as `true` in `opensearch.yml` (by default, both of these are `false`).
```

kolchfa-aws reviewed (approved):
Thank you, @noahstaveley! Some suggestions for you. Once addressed, I'll move the PR to editorial review.

natebower commented on _search-plugins/knn/knn-index.md:
```diff
@@ -60,14 +60,22 @@ SIMD optimization is applicable only if the vector dimension is a multiple of 8.
 <!-- vale on -->
 
-For the x64 architecture, two different versions of the Faiss library are built and shipped with the artifact:
+For the x64 architecture, the following versions of the Faiss library are built and shipped with the artifact:
```
```suggestion
For x64 architecture, the following versions of the Faiss library are built and shipped with the artifact:
```

natebower commented on _search-plugins/knn/knn-index.md:
```diff
@@ -60,14 +60,22 @@ SIMD optimization is applicable only if the vector dimension is a multiple of 8.
 
 - `libopensearchknn_faiss.so`: The non-optimized Faiss library without SIMD instructions. 
-- `libopensearchknn_faiss_avx2.so`: The Faiss library that contains AVX2 SIMD instructions. 
+- `libopensearchknn_faiss_avx512.so`: The Faiss library that contains AVX512 SIMD instructions. 
```
```suggestion
- `libopensearchknn_faiss_avx512.so`: The Faiss library containing AVX512 SIMD instructions. 
```

natebower commented on _search-plugins/knn/knn-index.md:
```diff
@@ -60,14 +60,22 @@ SIMD optimization is applicable only if the vector dimension is a multiple of 8.
 - `libopensearchknn_faiss.so`: The non-optimized Faiss library without SIMD instructions. 
-- `libopensearchknn_faiss_avx2.so`: The Faiss library that contains AVX2 SIMD instructions. 
+- `libopensearchknn_faiss_avx512.so`: The Faiss library that contains AVX512 SIMD instructions. 
+- `libopensearchknn_faiss_avx2.so`: The Faiss library that contains AVX2 SIMD instructions.
```
```suggestion
- `libopensearchknn_faiss_avx2.so`: The Faiss library containing AVX2 SIMD instructions.
```

natebower commented on _search-plugins/knn/knn-index.md:
```diff
@@ -60,14 +60,22 @@ SIMD optimization is applicable only if the vector dimension is a multiple of 8.
+- `libopensearchknn_faiss_avx2.so`: The Faiss library that contains AVX2 SIMD instructions.
 
-If your hardware supports AVX2, the k-NN plugin loads the `libopensearchknn_faiss_avx2.so` library at runtime.
+When using the Faiss library, the performance ranking is: AVX512 > AVX2 > no optimization.
```
```suggestion
When using the Faiss library, the performance ranking is as follows: AVX512 > AVX2 > no optimization.
```

natebower commented on _search-plugins/knn/settings.md:
```diff
@@ -27,6 +27,7 @@ Setting | Static/Dynamic | Default | Description
 `knn.model.index.number_of_replicas`| Dynamic | `1` | The number of replica shards to use for the model system index. Generally, in a multi-node cluster, this value should be at least 1 in order to increase stability.
 `knn.model.cache.size.limit` | Dynamic | `10%` |  The model cache limit cannot exceed 25% of the JVM heap.
 `knn.faiss.avx2.disabled` | Static | `false` | A static setting that specifies whether to disable the SIMD-based `libopensearchknn_faiss_avx2.so` library and load the non-optimized `libopensearchknn_faiss.so` library for the Faiss engine on machines with x64 architecture. For more information, see [SIMD optimization for the Faiss engine]({{site.url}}{{site.baseurl}}/search-plugins/knn/knn-index/# ...
+`knn.faiss.avx512.disabled` | Static | `false` | A static setting that specifies whether to disable the SIMD-based `libopensearchknn_faiss_avx512.so` library and load the `libopensearchknn_faiss_avx2.so` or the non-optimized `libopensearchknn_faiss.so` library for the Faiss engine on machines with x64 architecture. For more information, see [SIMD optimization for the Faiss engine]({{site.url}}{{s ...
```
```suggestion
`knn.faiss.avx512.disabled` | Static | `false` | A static setting that specifies whether to disable the SIMD-based `libopensearchknn_faiss_avx512.so` library and load the `libopensearchknn_faiss_avx2.so` library or the non-optimized `libopensearchknn_faiss.so` library for the Faiss engine on machines with x64 architecture. For more information, see [SIMD optimization for the Faiss engine]({{site.url}}{{site.baseurl}}/search-plugins/knn/knn-index/#simd-optimization-for-the-faiss-engine).
```
